### PR TITLE
Install Java 8, Android command line tools and SDK for Ubuntu

### PR DESCRIPTION
**Pull request by jpsim**
These are used for Envoy Mobile's CI and currently installed on every CI job run: https://github.com/envoyproxy/envoy/blob/main/mobile/ci/linux_ci_setup.sh

### CODE REVIEW

**Review by phlax — COMMENTED** [review]
im wondering if it would be possible to do this with bazel - seems like there are bazel rules - https://bazel.build/docs/android-ndk

**Comment by jpsim** [comment]
I think Keith looked into it and ran into this issue: https://github.com/bazelbuild/rules_android_ndk/pull/24

File: build_container/build_container_ubuntu.sh [hunk 1]
```diff
@@ -132,6 +132,60 @@ update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 1
...
+    mv cmdline-tools/* sdk/cmdline-tools/latest
+
+    ANDROID_HOME="$(realpath "$sdk_install_target/sdk")"
+    SDKMANAGER=$ANDROID_HOME/cmdline-tools/latest/bin/sdkmanager
+
+    echo "y" | $SDKMANAGER --install "ndk;21.4.7075529"
+    $SDKMANAGER --install "platforms;android-30"
+    $SDKMANAGER --install "build-tools;30.0.2"
+
+    export "ANDROID_HOME=${ANDROID_HOME}"
```
Comment (phlax) [inline build_container/build_container_ubuntu.sh:176]: i think these wont be available in the container env when it is run and dont seem to be used in the build

Comment (jpsim) [reply build_container/build_container_ubuntu.sh:176]: I’ll look into how to set env variables in Docker from a bash script then.

the intention is to have these set when the container runs.

Comment (phlax) [reply build_container/build_container_ubuntu.sh:176]: i think you want to set them as `ENV` vars in the `Dockerfile` and then you can use them here and at runtime

Comment (jpsim) [reply build_container/build_container_ubuntu.sh]: Done in 6f3544ff5b8d0d47c4c234f9e1842e97b8243810.

File: build_container/build_container_ubuntu.sh [hunk 2]
```diff
@@ -132,6 +132,55 @@ update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 1
...
+    apt-get update -y
+    apt-get install -y zulu8-jdk
+    rm ./zulu-repo_1.0.0-3_all.deb
+
+    #######################
+    # Install Android tools
+    #######################
+
+    sdk_install_target="/.android"
+    android_home="$sdk_install_target/sdk"
```
Comment (phlax) [inline build_container/build_container_ubuntu.sh:160]: i think rather use the vars from the env (and i guess add either add `$ANDROID_INSTALL_TARGET` or set the `sdk_install_target` as parent of `$ANDROID_HOME`)

it just means that its set one place authoritatively and less risk of it getting out of sync in some way

File: build_container/build_container_ubuntu.sh [hunk 3]
```diff
@@ -132,6 +132,55 @@ update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.10 1
...
+    pushd "$sdk_install_target"
+
+    cmdline_file="commandlinetools-linux-7583922_latest.zip"
+    curl -OL "https://dl.google.com/android/repository/$cmdline_file"
+    unzip "$cmdline_file"
+    mkdir -p sdk/cmdline-tools/latest
+    mv cmdline-tools/* sdk/cmdline-tools/latest
+
+    sdkmanager=$android_home/cmdline-tools/latest/bin/sdkmanager
+    echo "y" | $sdkmanager --install "ndk;21.4.7075529"
```
Comment (phlax) [inline build_container/build_container_ubuntu.sh:171]: im wondering similar about the version here as its ~set as an env var also

File: build_container/build_container_ubuntu.sh (hunk 2 above)
Comment (jpsim) [reply build_container/build_container_ubuntu.sh]: Please let me know if I misunderstood your request, I tried to do what you were suggesting in 1a127b2399beae4181e5594f9d34b08ba36f1063.

**Review by phlax — APPROVED** [review]
thanks for addressing @jpsim , lgtm

**Comment by jpsim** [comment]
This grew the size of the Docker image by quite a bit: 1.79 GB to 3.77 GB

That's unfortunate given the vast majority of Envoy's CI jobs don't need the Android toolchain installed.

@phlax what if we added a new image that was Ubuntu+Android so that most CI jobs can use the leaner image, and the Envoy Mobile CI jobs can use this heavier one?

I'm also seeing some issues even using this image on CI: https://github.com/envoyproxy/envoy/pull/24460

**Comment by phlax** [comment]
> This grew the size of the Docker image by quite a bit: 1.79 GB to 3.77 GB

that was my original concern - that is going to be a huge problem - it will slow all envoy ci down a lot (3.77 is the compressed size to in reality it will take a lot more space)

> I'm also seeing some issues even using this image on CI: https://github.com/envoyproxy/envoy/pull/24460

yep we already overload dockerhub and have intermittent issues

> @phlax what if we added a new image that was Ubuntu+Android so that most CI jobs can use the leaner image, and the Envoy Mobile CI jobs can use this heavier one?

i really dont see any other way - it will take a bit to set up cc @lizan

**Comment by phlax** [comment]
i guess its worth asking my original question again - is there no way we could do this with bazel ?

**Comment by jpsim** [comment]
I'll revert this until we sort out of a better solution.

> is there no way we could do this with bazel ?

I'd love to find a way to do this with Bazel, but I think the blocker is still here: https://github.com/bazelbuild/rules_android_ndk/pull/24

**Comment by phlax** [comment]
looking at the fail trying to use this - i think even if we separated out to a different image it will still have issues in github actions being too large

**Comment by jpsim** [comment]
> i think even if we separated out to a different image it will still have issues in github actions being too large

The issue isn't on the GitHub Actions side, it's on the EngFlow RBE side. It's being looked into here: https://envoyproxy.slack.com/archives/C02QMNG92A3/p1670596188932999

**Pull request merged**